rust/experimental/server/src/store/memory.rs

```diff
@@ -343,21 +345,45 @@ impl Store for MemoryStore {
             .instrument_await("getting partitioned buffer lock")
             .await;
 
+        // get block_ids filter
+        // In AQE, after executing the sub-QueryStages, collect the shuffle data size
+        // So if we can filter block, it will improve the performance of AQE.
+        let block_ids_filter = if !ctx.serialized_expected_task_ids_bitmap.is_empty() {
+            match Treemap::deserialize(&ctx.serialized_expected_task_ids_bitmap) {
+                Ok(filter) => Some(filter),
+                Err(e) => {
+                    error!("Failed to deserialize: {}", e);
+                    None
+                }
+            }
+        } else {
+            None
+        };
+
         let options = ctx.reading_options;
         let (fetched_blocks, length) = match options {
             MEMORY_LAST_BLOCK_ID_AND_MAX_SIZE(last_block_id, max_size) => {
                 let mut last_block_id = last_block_id.clone();
                 let mut in_flight_flatten_blocks = vec![];
                 for (_, blocks) in buffer.in_flight.iter() {
                     for in_flight_block in blocks {
-                        in_flight_flatten_blocks.push(in_flight_block);
+                        if block_ids_filter.as_ref().map_or(true, |filter| {
```

Comment:
This part is to group the in_flight + staging data, which is not suitable for filtering. The filtering should be operated after the `last_block_id` is found.